File: ant-sdk/antd-php/src/DaemonDiscovery.php
<?php

declare(strict_types=1);

namespace Autonomi\Antd;

/**
 * Discovers the antd daemon by reading the `daemon.port` file written on startup.
 *
 * The port file contains up to three lines: REST port (line 1), gRPC port (line 2),
 * and PID of the daemon process (line 3). If a PID is present and the process is
 * not alive, the file is considered stale and discovery returns empty.
 * File location is platform-specific:
 * - Windows: %APPDATA%\ant\sdk\daemon.port
 * - macOS:   ~/Library/Application Support/ant/sdk/daemon.port
 * - Linux:   $XDG_DATA_HOME/ant/sdk/daemon.port or ~/.local/share/ant/sdk/daemon.port
 *
 * The `sdk` subdirectory keeps antd's port file separate from the ant-node
 * daemon, which writes to the same `ant` umbrella dir.
 */
class DaemonDiscovery
{
    private const PORT_FILE_NAME = 'daemon.port';
    private const DATA_DIR_NAME = 'ant';
    private const SDK_SUBDIR_NAME = 'sdk';

    /**
     * Reads the daemon.port file and returns the REST base URL
     * (e.g. "http://127.0.0.1:8082"). Returns "" if unavailable.
     */
    public static function discoverDaemonUrl(): string
    {
        [$rest] = self::readPortFile();
        if ($rest === 0) {
            return '';
        }
        return "http://127.0.0.1:{$rest}";
    }

    /**
     * Reads the daemon.port file and returns the gRPC target
     * (e.g. "127.0.0.1:50051"). Returns "" if unavailable.
     */
    public static function discoverGrpcTarget(): string
    {
        [, $grpc] = self::readPortFile();
        if ($grpc === 0) {
            return '';
        }
        return "127.0.0.1:{$grpc}";
    }

    /**
     * Create an AntdClient using the discovered daemon URL.
     * Falls back to http://localhost:8082 if discovery fails.
     *
     * @param float $timeout Request timeout in seconds.
     * @param \GuzzleHttp\Client|null $httpClient Optional HTTP client.
     * @return array{0: AntdClient, 1: string} [$client, $url]
     */
    public static function autoDiscover(float $timeout = 300.0, ?\GuzzleHttp\Client $httpClient = null): array
    {
        $url = self::discoverDaemonUrl();
        if ($url === '') {
            $url = 'http://localhost:8082';
        }
        $client = new AntdClient($url, $timeout, $httpClient);
        return [$client, $url];
    }

    /**
     * @return array{0: int, 1: int} [restPort, grpcPort]
     */
    private static function readPortFile(): array
    {
        $dir = self::dataDir();
        if ($dir === '') {
            return [0, 0];
        }

        $path = $dir . DIRECTORY_SEPARATOR . self::PORT_FILE_NAME;
        $contents = @file_get_contents($path);
        if ($contents === false) {
            return [0, 0];
        }

        $lines = explode("\n", trim($contents));
        if (count($lines) < 1) {
            return [0, 0];
        }

        // Line 3: PID of the daemon process (optional stale-detection)
        if (count($lines) >= 3) {
            $pid = (int) trim($lines[2]);
            if ($pid > 0 && !self::isProcessAlive($pid)) {
                return [0, 0];
            }
        }

        $rest = self::parsePort($lines[0]);
        $grpc = count($lines) >= 2 ? self::parsePort($lines[1]) : 0;
        return [$rest, $grpc];
    }

    /**
     * Check if a process with the given PID is alive.
     * Uses posix_kill if available, falls back to /proc on Linux,
     * or tasklist on Windows.
     */
    private static function isProcessAlive(int $pid): bool
    {
        if ($pid <= 0) {
            return false;
        }

        if (PHP_OS_FAMILY === 'Windows') {
            // Use COM WMI query to avoid shell_exec injection risks.
            // Falls back to trusting the port file if COM is unavailable.
            if (class_exists('COM', false)) {
                try {
                    /** @phpstan-ignore-next-line */
                    $wmi = new \COM('WbemScripting.SWbemLocator');
                    /** @phpstan-ignore-next-line */
                    $service = $wmi->ConnectServer('.', 'root\\cimv2');
                    $query = $service->ExecQuery(
                        'SELECT ProcessId FROM Win32_Process WHERE ProcessId = ' . $pid
                    );
                    return $query->Count > 0;
                } catch (\Throwable) {
                    return true; // Cannot check; trust the port file.
                }
            }
            return true; // COM not available; trust the port file.
        }

        // Unix: prefer posix_kill, fall back to /proc
        if (function_exists('posix_kill')) {
            return posix_kill($pid, 0);
        }

        return file_exists("/proc/{$pid}");
    }

    private static function parsePort(string $s): int
    {
        $n = (int) trim($s);
        if ($n < 1 || $n > 65535) {
            return 0;
        }
        return $n;
    }

    private static function dataDir(): string
    {
        switch (PHP_OS_FAMILY) {
            case 'Windows':
                $appdata = getenv('APPDATA');
                if ($appdata === false || $appdata === '') {
                    return '';
                }
                return $appdata . DIRECTORY_SEPARATOR . self::DATA_DIR_NAME . DIRECTORY_SEPARATOR . self::SDK_SUBDIR_NAME;

            case 'Darwin':
                $home = getenv('HOME');
                if ($home === false || $home === '') {
                    return '';
                }
                return $home . '/Library/Application Support/' . self::DATA_DIR_NAME . '/' . self::SDK_SUBDIR_NAME;

            default: // Linux and others
                $xdg = getenv('XDG_DATA_HOME');
                if ($xdg !== false && $xdg !== '') {
                    return $xdg . '/' . self::DATA_DIR_NAME . '/' . self::SDK_SUBDIR_NAME;
                }
                $home = getenv('HOME');
                if ($home === false || $home === '') {
                    return '';
                }
                return $home . '/.local/share/' . self::DATA_DIR_NAME . '/' . self::SDK_SUBDIR_NAME;
        }
    }
}

File: ant-sdk/antd-csharp/Antd.Sdk/DaemonDiscovery.cs
using System.Runtime.InteropServices;

namespace Antd.Sdk;

/// <summary>
/// Reads the daemon.port file written by antd on startup to auto-discover
/// the REST and gRPC ports. The file contains up to three lines: REST port
/// on line 1, gRPC port on line 2, and daemon PID on line 3. If a PID is
/// present and the process is no longer alive, the file is considered stale
/// and discovery returns empty.
/// </summary>
public static class DaemonDiscovery
{
    private const string PortFileName = "daemon.port";
    private const string DataDirName = "ant";
    private const string SdkSubDirName = "sdk";

    /// <summary>
    /// Reads line 1 of the daemon.port file and returns the REST base URL
    /// (e.g. "http://127.0.0.1:8082"). Returns empty string on failure.
    /// </summary>
    public static string DiscoverDaemonUrl()
    {
        var (restPort, _) = ReadPortFile();
        return restPort > 0 ? $"http://127.0.0.1:{restPort}" : "";
    }

    /// <summary>
    /// Reads line 2 of the daemon.port file and returns the gRPC target
    /// (e.g. "http://127.0.0.1:50051"). Returns empty string on failure.
    /// </summary>
    public static string DiscoverGrpcTarget()
    {
        var (_, grpcPort) = ReadPortFile();
        return grpcPort > 0 ? $"http://127.0.0.1:{grpcPort}" : "";
    }

    private static (ushort restPort, ushort grpcPort) ReadPortFile()
    {
        var dir = DataDir();
        if (string.IsNullOrEmpty(dir))
            return (0, 0);

        var path = Path.Combine(dir, PortFileName);
        if (!File.Exists(path))
            return (0, 0);

        try
        {
            var text = File.ReadAllText(path).Trim();
            var lines = text.Split('\n');

            ushort rest = 0, grpc = 0;
            if (lines.Length >= 1)
                rest = ParsePort(lines[0]);
            if (lines.Length >= 2)
                grpc = ParsePort(lines[1]);

            // Line 3 is the daemon PID. If present and the process is
            // no longer running, the port file is stale.
            if (lines.Length >= 3
                && int.TryParse(lines[2].Trim(), out var pid)
                && pid > 0
                && !ProcessAlive(pid))
            {
                return (0, 0);
            }

            return (rest, grpc);
        }
        catch
        {
            return (0, 0);
        }
    }

    private static bool ProcessAlive(int pid)
    {
        try
        {
            System.Diagnostics.Process.GetProcessById(pid);
            return true;
        }
        catch (ArgumentException)
        {
            return false;
        }
    }

    private static ushort ParsePort(string s)
    {
        return ushort.TryParse(s.Trim(), out var port) ? port : (ushort)0;
    }

    /// <summary>
    /// Returns the platform-specific data directory for the antd SDK daemon.
    ///   Windows: %APPDATA%\ant\sdk
    ///   macOS:   ~/Library/Application Support/ant/sdk
    ///   Linux:   $XDG_DATA_HOME/ant/sdk or ~/.local/share/ant/sdk
    ///
    /// The "sdk" subdirectory keeps antd's port file separate from the
    /// ant-node daemon, which writes to the same "ant" umbrella dir.
    /// </summary>
    private static string DataDir()
    {
        if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
        {
            var appdata = Environment.GetEnvironmentVariable("APPDATA");
            return string.IsNullOrEmpty(appdata) ? "" : Path.Combine(appdata, DataDirName, SdkSubDirName);
        }

        if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
        {
            var home = Environment.GetEnvironmentVariable("HOME");
            return string.IsNullOrEmpty(home) ? "" : Path.Combine(home, "Library", "Application Support", DataDirName, SdkSubDirName);
        }

        // Linux and others
        var xdg = Environment.GetEnvironmentVariable("XDG_DATA_HOME");
        if (!string.IsNullOrEmpty(xdg))
            return Path.Combine(xdg, DataDirName, SdkSubDirName);

        var homeDir = Environment.GetEnvironmentVariable("HOME");
        return string.IsNullOrEmpty(homeDir) ? "" : Path.Combine(homeDir, ".local", "share", DataDirName, SdkSubDirName);
    }
}
